### PR TITLE
Support ILQL for T5 model, Fix PPO T5 for refactored code

### PR DESCRIPTION
**Pull request by PhungVanDuy**
Adapted ILQL implementation for Causal models, this PR support ILQL for the T5 model. 
- Related issue: https://github.com/CarperAI/trlx/issues/204

- [x] Add ILQL for the T5 model
- [x] Run Summarization example by OpenAI dataset
- [x] Add sentiment examples for PPO T5
- [x] Add tests

ILQL T5 on OpenAI Dataset: https://wandb.ai/pvduy/trlx/runs/sm5gug89
Sentiment T5: https://wandb.ai/pvduy/trlx/runs/vgvph8cc

### CODE REVIEW

File: configs/ilql_summarize_t5.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  checkpoint_interval: 100000
+  eval_interval: 500
+
+  pipeline: "PromptPipeline"
+  trainer: "AccelerateILQLTrainer"
+  checkpoint_dir: "gpt-j-ilql-summarize-duy"
+  #  tracker: null
+
+model:
+  model_path: "/admin/home-duyphung/software/github/pvd/trlx/examples/summarize_rlhf/flan-t5-qa/checkpoint-3000"
```
Comment (reciprocated) [inline configs/ilql_summarize_t5.yml:16]: this path is unfortunately private, but you don't have to use sft and can just start from hf's checkpoint here

File: examples/summarize_rlhf/ilql_summarize_t5.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            )
+            input_ids = encodings_dict["input_ids"].to(rw_device)
+            attn_masks = encodings_dict["attention_mask"].to(rw_device)
+            input_ids = input_ids.repeat(2, 1)
+            attn_masks = attn_masks.repeat(2, 1)
+            with torch.no_grad():
+                sub_scores = rw_model(input_ids=input_ids, attention_mask=attn_masks)
+            scores_list.append(sub_scores["chosen_end_scores"])
+        scores = torch.cat(scores_list, dim=0)
+        return scores
```
Comment (reciprocated) [inline examples/summarize_rlhf/ilql_summarize_t5.py:58]: maybe we can reshare code for the reward model instantiation with ppo_t5 as well

File: trlx/orchestrator/offline_orchestrator.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
+import os
```
Comment (reciprocated) [inline trlx/orchestrator/offline_orchestrator.py:1]: the contents of this file were refactored back into `accelerate_ilql_trainer.py` and the file itself was deleted

File: trlx/trainer/nn/ilql_models.py [hunk 4]
```diff
@@ -62,12 +63,15 @@ class ILQLConfig(MethodConfig):
...
         logits, (qs, target_qs, vs) = outputs
         terminal_mask = labels.dones[:, :-1]
         n_nonterminal = max(1, terminal_mask.sum())
-
-        actions = labels.input_ids[:, 1:].gather(dim=1, index=labels.actions_ixs).unsqueeze(-1)
+        # check type of labels
+        if isinstance(labels, ILQLBatch):
+            actions = labels.input_ids[:, 1:].gather(dim=1, index=labels.actions_ixs).unsqueeze(-1)
+        else:
+            actions = labels.decoder_input_ids[:, 1:].gather(dim=1, index=labels.actions_ixs).unsqueeze(-1)
```
Comment (reciprocated) [inline trlx/trainer/nn/ilql_models.py:74]: why do we want to `gather` here, wouldn't `decoder_input_ids` be already compiled in the needed format from `make_experience`

File: trlx/trainer/nn/ilql_models.py [hunk 5]
```diff
@@ -331,3 +335,144 @@ def dummy_inputs(self):
...
+        Generates samples akin to hf's `.generate` but with custom logp prepossessing:
+        changing token probabilities as to how advantageous they would be
+        according to value functions estimations.
+        """
+        if attention_mask is None:
+            attention_mask = input_ids.not_equal(pad_token_id)
+        samples = input_ids.clone()
+        max_new_tokens = min(max_new_tokens, max_length - input_ids.shape[1])
+        if decoder_input_ids is None:
+            decoder_input_ids = input_ids.new_zeros(input_ids.shape[0], 1, dtype=torch.long, device=input_ids.device)
```
Comment (reciprocated) [inline trlx/trainer/nn/ilql_models.py:436]: it's either `torch.zeros` or remove `dtype` and `device` in call to `new_zeros`

File: trlx/trainer/nn/ilql_models.py [hunk 6]
```diff
@@ -331,3 +335,144 @@ def dummy_inputs(self):
...
+            adv = qs - vs
+            pi_beta = F.log_softmax(logits, -1)
+            pi_top_k = topk_mask(pi_beta + beta * adv, top_k)
+            pi = F.softmax(pi_top_k / temperature, -1)
+            next_tokens = torch.multinomial(pi, num_samples=1)
+            if eos_token_id is not None:
+                next_tokens = (1 - finished) * next_tokens + finished * pad_token_id
+                finished = (next_tokens == eos_token_id).long()
+            else:
+                raise ValueError("eos_token_id must be provided")
```
Comment (reciprocated) [inline trlx/trainer/nn/ilql_models.py:463]: this check is great, however it's better to do it at the start of the function and maybe not in the loop

**Review by reciprocated — COMMENTED** [review]
Glad that it works overall, but we still have to find a way to refactor many repetitions here, at least up to the same level as it's done in PPO

File: trlx/trainer/nn/ilql_models.py (hunk 4 above)
Comment (PhungVanDuy) [reply trlx/trainer/nn/ilql_models.py:74]: yes, you are right we do no need the gather here. Actions also started with 0. Will remove it, but it not effect current result.

File: configs/ilql_summarize_t5.yml (hunk 1 above)
Comment (PhungVanDuy) [reply configs/ilql_summarize_t5.yml:16]: Yes I will upload T5 SFT for to hf.

File: trlx/trainer/accelerate_ilql_trainer.py [hunk 7]
```diff
@@ -80,12 +99,98 @@ def prepare_learning(self):
...
+            logger.info("Logging experience string statistics")
+            columns = ["Prompt Length", "Output Length", "Sample Length"]
+            table = Table(*columns, title="Experience String Stats (mean ∈ \[min, max])", show_lines=True)
+            row = []
+            for lengths in [prompt_lengths, output_lengths, sample_lengths]:
+                row.append(f"{lengths.mean():.2f} ∈ [{min(lengths)}, {max(lengths)}]")
+            table.add_row(*row)
+            Console().print(table)
+
+        returns = (returns - returns.mean()) / (returns.std() + 1e-30)
```
Comment (jon-tow) [inline trlx/trainer/accelerate_ilql_trainer.py:169]: Nit: We probably shouldn't hardcode the epsilon like this at 1e-30 and just use the machine epsilon, for example`torch.finfo(returns.dtype).eps`

File: trlx/trainer/accelerate_ilql_trainer.py [hunk 8]
```diff
@@ -80,12 +99,98 @@ def prepare_learning(self):
...
+            all_actions_ixs,
+            all_dones,
+        )
+
+    def make_experience(self, samples, rewards, max_length=2048):
+        """
+        Tokenizes samples and shapes rewards into proper tensors and then inserts the resulting dataset into the trainer
+        """
+
+        if self.config.model.model_arch_type == "seq2seq":
```
Comment (jon-tow) [inline trlx/trainer/accelerate_ilql_trainer.py:190]: Aside: At some point, we need to design a better way to handle these arch-type conditionals 😅

File: trlx/models/modeling_ilql.py [hunk 9]
```diff
@@ -339,3 +353,147 @@ def post_init(self, state_dict):
         self.ilql_heads.load_state_dict(state_dict, strict=False)
         del state_dict
         gc.collect()
+
+
+class AutoModelForSeq2SeqLMWithILQLHeads(PreTrainedModelWrapper):
```
Comment (jon-tow) [inline trlx/models/modeling_ilql.py:358]: Let's add a test for this similar to https://github.com/CarperAI/trlx/blob/e72f7d1a8008c9a994e9fe465aa4a8a7a1fb3232/tests/test_models.py#L138

File: trlx/trainer/accelerate_ppo_trainer.py [hunk 10]
```diff
@@ -170,7 +174,7 @@ def loss(self, batch: PPORLBatch):
             end = start + response_length
             logprobs, values_pred, mask = (
                 logprobs[:, start:end],
-                values_pred[:, start - 1 : end - 1],
+                values_pred[:, start:end],
```
Comment (jon-tow) [inline trlx/trainer/accelerate_ppo_trainer.py:177]: We need to add a test for this

**Review by jon-tow — CHANGES_REQUESTED** [review]
Hi, @PhungVanDuy, great work! I've left some feedback to look at when you get a chance 👍  👍

File: trlx/data/ilql_types.py [hunk 11]
```diff
@@ -38,6 +38,30 @@ class ILQLElement:
...
+    :param attention_mask: Attention mask. Should be a long tensor.
+    :type attention_mask: torch.Tensor
+
+    :param rewards: Rewards for each token. Should be a float tensor of same size as tokens.
+    :type rewards: torch.Tensor
+    """
+
+    input_ids: TensorType["query_size"]
+    attention_mask: TensorType["query_size"]
+    decoder_input_ids: TensorType["reward_size"]
```
[the remark this answers is not in the document]
Comment (PhungVanDuy) [reply trlx/data/ilql_types.py:58]: I prefer to separate this one for easier maintenance. We can re-design this one along when we find a way to seperate "seq2seq" like you said here: https://github.com/CarperAI/trlx/pull/290#discussion_r1122524562

**Review by jon-tow — APPROVED** [review]
Looks good on my end! Great work :)

**Comment by PhungVanDuy** [comment]
> Looks good on my end! Great work :)

Thank you so much for your comment!
